### PR TITLE
feat: adjustment invoice

### PR DESCRIPTION
**Pull request by ribbery009**

### CODE REVIEW

File: lib/Invoice.js [hunk 1]
```diff
@@ -66,6 +68,8 @@ export class Invoice {
     assert(Array.isArray(this._options.items),
       'Valid Items array missing from invoice options')
 
+    assert(!(this._options.adjustmentInvoice === true && typeof this._options.adjustmentInvoiceNumber === 'undefined'), 'If adjustmentInvoice is true, adjustmentInvoiceNumber must be specified.');
```
Comment (ert78gb) [inline lib/Invoice.js:71]: Unfortunately the [documentation](https://docs.szamlazz.hu/#xsd-scheme-compliance) does not really describe enough details about these fields. 

For me, it is more logical if we keep this package as simple as possible. What do you think if we introduce only the `adjustmentInvoiceNumber` as a constructor parameter. And if it is a valid string we set the `this._options.adjustmentInvoice` true. In this case the developers do not have to set both fields.

I think the following logic is more proper
```javascript
if (this._options.adjustmentInvoiceNumber !== null && this._options.adjustmentInvoiceNumber !== undefined) {
  assert(typeof this._options.adjustmentInvoiceNumber !== 'string', '"adjustmentInvoiceNumber" should be a string')
  assert(this._options.adjustmentInvoiceNumber.length > 0, '"adjustmentInvoiceNumber" should be minimum 1 character')

  this._options.adjustmentInvoice = true
}
```

Please also write tests and extend the `readme.md` too.

Comment (ribbery009) [reply lib/Invoice.js]: Unfortunately, this part of the documentation is not clearly defined. Thank you for the example correction; it was a great idea. I made only this change:

assert(typeof` this._options.adjustmentInvoiceNumber === 'string', '"adjustmentInvoiceNumber" should be a `string')

I've completed the tests, tried them manually with invoices as well, and the cases were successful. Additionally, I've updated the README.md file. Is this acceptable, or should I change anything?

Comment (ert78gb) [reply lib/Invoice.js]: Please add unit tests to the `tests/invoice.spec.js`. The following tests are needed
- `adjustmentInvoiceNumber` is `null`
- `adjustmentInvoiceNumber` is `undefined`
- `adjustmentInvoiceNumber` is empty `string`
- `adjustmentInvoiceNumber` is not `string`. You can test it with `Date`, `Number` or `Boolean` or with all the 3 :)
- `adjustmentInvoiceNumber` is a non empty `string`

I think, good idea to create a new test suite `describe` block in the `_generateXML` test suite.

Comment (ribbery009) [reply lib/Invoice.js]: As requested, I added the unit tests to the tests/invoice.spec.js file. The tests cover the following scenarios:

- adjustmentInvoiceNumber null
- adjustmentInvoiceNumber is undefined
- adjustmentInvoiceNumber empty string
- adjustmentInvoiceNumber is not a string. The tests also apply to Date, Number and Boolean types.
- adjustmentInvoiceNumber is not an empty string

As suggested, I have created a new test suite within the _generateXML test suite that contains these tests.

Thanks very much for your guidance and patience; it was very helpful and I managed to implement the tests accordingly. 
What do you think, is this suitable?

File: tests/invoice.spec.js [hunk 2]
```diff
@@ -72,6 +72,60 @@ describe('Invoice', function () {
...
+            expect(result.wrapper).to.not.have.deep.property('fejlec.helyesbitettSzamlaszam');
+            done();
+          });
+        });
+
+        it('should throw an error when adjustmentInvoiceNumber is an empty string', function () {
+          expect(() => {
+            invoice._options.adjustmentInvoiceNumber = '';
+            invoice._generateXML();
+          }).to.throw();
```
Comment (ert78gb) [inline tests/invoice.spec.js:98]:
```suggestion
          }).to.throw(/"adjustmentInvoiceNumber" should be minimum 1 character/);
```

File: tests/invoice.spec.js [hunk 3]
```diff
@@ -72,6 +72,60 @@ describe('Invoice', function () {
...
+        });
+
+        it('should throw an error when adjustmentInvoiceNumber is not a string', function () {
+          const invalidTypes = [new Date(), 123, true];
+          invalidTypes.forEach(type => {
+            expect(() => {
+              invoice._options.adjustmentInvoiceNumber = type;
+              invoice._generateXML();
+            }).to.throw();
```
Comment (ert78gb) [inline tests/invoice.spec.js:107]:
```suggestion
            }).to.throw(/"adjustmentInvoiceNumber" should be a string/);
```

File: tests/invoice.spec.js [hunk 4]
```diff
@@ -72,6 +72,60 @@ describe('Invoice', function () {
...
+
+        it('should throw an error when adjustmentInvoiceNumber is an empty string', function () {
+          expect(() => {
+            invoice._options.adjustmentInvoiceNumber = '';
+            invoice._generateXML();
+          }).to.throw();
+        });
+
+        it('should throw an error when adjustmentInvoiceNumber is not a string', function () {
+          const invalidTypes = [new Date(), 123, true];
```
Comment (ert78gb) [inline tests/invoice.spec.js:102]: it is nice simplification, but if it fails not easy to debug what was wrong.
There are 2 options.

Dynamically create test `it`
```javascript
const invalidTypes = [new Date(), 123, true];
invalidTypes.forEach(type => {
  it(`should throw an error when adjustmentInvoiceNumber is ${typeof type}`, function () {
    expect(() => {
    invoice._options.adjustmentInvoiceNumber = type;
    invoice._generateXML();
    }).to.throw(/"adjustmentInvoiceNumber" should be a string/);
  });
});
```

Create 3 independent test

My problem with the 1st approach is
- `typeof new Date` is `object`. That logically good, but 
- you can't test/debug only 1 test if fail.
- I don't like overcomplicate test. It should have minimal logic. I also don't like code duplication but code duplication in test are not same as in code. It is my personal preference, but many times when the code changed I did not have to rewrite the "testing world"

So I vote on the 2nd solution, but I accept the dynamic version too.

**Review by ert78gb — CHANGES_REQUESTED** [review]

File: tests/invoice.spec.js (hunk 4 above)
Comment (ribbery009) [reply tests/invoice.spec.js]: Thank you very much for your time and the valuable suggestions for testing. It was great consulting and discussing these matters with you. I found it very beneficial and I agree with your recommendations. I have endeavored to implement them as expected.

**Review by ert78gb — APPROVED** [review]

**Comment by ert78gb** [comment]
thanks for the PR.

**Pull request merged**